march_hardware/src/PDOmap.cpp

```diff
@@ -0,0 +1,166 @@
+{
+  this->sortPDOObjects();
+  int reg;
+  int SMAddress;
+  if (direction == dataDirection::miso)
+  {
+    reg = 0x1A00;
+    SMAddress = 0x1C13;
+  }
+  else if (direction == dataDirection::mosi)
+  {
+    reg = 0x1600;
+    SMAddress = 0x1C12;
+  }
+  else
+  {
+    ROS_ERROR("Invalid dataDirection argument");
+  }
+  // Clear SyncManager Object
+  sdo_bit8(slaveIndex, SMAddress, 0, 0);
+  ROS_DEBUG("sdo write: slaveindex %i, reg 0x%X, subindex 0, value 0x0", slaveIndex, SMAddress);
+  int startReg = reg;
+  int lastFilledReg = reg;
+  int sizeleft = this->bitsPerReg;
+  int counter = 0;
+  int byteOffset = 0;
+  while (this->sortedPDOObjects.size() > 0)
+  {
+    // Check if register is still empty
+    if (sizeleft == this->bitsPerReg)
+    {
+      sdo_bit32(slaveIndex, reg, 0, 0);
+      ROS_DEBUG("sdo write: slaveIndex %i, reg 0x%X, subindex 0, value 0x0", slaveIndex, reg);
+    }
+    // Get next object (from end, because sorted from small to large)
+    std::pair<IMCObjectName, IMCObject> nextObject = this->sortedPDOObjects.back();
+    this->sortedPDOObjects.pop_back();
+    // Add next object to map
+    counter++;
+    sdo_bit32(slaveIndex, reg, counter,
+              this->combineAddressLength(nextObject.second.address, nextObject.second.length));
+    ROS_DEBUG("sdo write: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slaveIndex, reg, counter,
+              this->combineAddressLength(nextObject.second.address, nextObject.second.length));
+    this->byteOffsets[nextObject.first] = byteOffset;
+    byteOffset += nextObject.second.length / 8;
+    sizeleft -= nextObject.second.length;
+    // Check if this was the last object of the list
+    if (this->sortedPDOObjects.size() == 0)
+    {
+      sdo_bit32(slaveIndex, reg, 0, counter);
+      ROS_DEBUG("sdo write: slaveIndex %i, reg 0x%X, subindex 0, value 0x%X", slaveIndex, reg, counter);
+      lastFilledReg = reg;
+      reg++;
+    }
+    // else, check if register is full
+    else if (sizeleft <= 0)
+    {
+      sdo_bit32(slaveIndex, reg, 0, counter);
+      ROS_DEBUG("sdo write: slaveIndex %i, reg 0x%X, subindex 0, value 0x%X", slaveIndex, reg, counter);
+      reg++;
+      counter = 0;
+      sizeleft = this->bitsPerReg;
+    }
+  }
+  // For the unused registers, set count to zero
+  for (int i = reg; i < startReg + this->nrofRegs; i++)
+  {
+    sdo_bit32(slaveIndex, i, 0, 0);
+    ROS_DEBUG("sdo write: slaveIndex %i, reg 0x%X, subindex 0, value 0x0", slaveIndex, i);
+  }
+  // For all filled registers, set data to Sync Manager object
+  int count = 0;
+  for (int i = startReg; i <= lastFilledReg; i++)
+  {
+    count++;
+    sdo_bit16(slaveIndex, SMAddress, count, 0x1600);
+    ROS_DEBUG("sdo write: slaveindex %i, reg 0x%X, subindex %i, value 0x%X", slaveIndex, SMAddress, count, i);
+  }
+  sdo_bit8(slaveIndex, SMAddress, 0, count);
+  ROS_DEBUG("sdo write: slaveindex %i, reg 0x%X, subindex 0, value 0x%X", slaveIndex, SMAddress, count);
```

Comment:
Place debugs before the function incase it throws an exception

Reply:
Debugs prints now inside sdo_bit functions, before executing